packages/client/src/rtc/publisher.ts

```diff
@@ -92,64 +102,69 @@ export class Publisher {
           t.sender.track?.kind === this.trackKindRegistry[trackType],
       );
 
+    /**
+     * An event handler which listens for the 'ended' event on the track.
+     * Once the track has ended, it will notify the SFU and update the state.
+     */
+    const handleTrackEnded = async () => {
+      console.log(`Track ${TrackType[trackType]} has ended, notifying the SFU`);
+      await this.notifyTrackMuteStateChanged(mediaStream, trackType, true);
+      // clean-up, this event listener needs to run only once.
+      track.removeEventListener('ended', handleTrackEnded);
+    };
+
     if (!transceiver) {
-      let videoEncodings: RTCRtpEncodingParameters[] | undefined;
-      if (trackType === TrackType.VIDEO) {
-        videoEncodings = findOptimalVideoLayers(track);
-      }
+      const videoEncodings =
+        trackType === TrackType.VIDEO
+          ? findOptimalVideoLayers(track)
+          : undefined;
+
+      const codecPreferences = this.getCodecPreferences(
+        trackType,
+        opts.preferredCodec,
+      );
+
+      // listen for 'ended' event on the track as it might be ended abruptly
+      // by an external factor as permission revokes, device disconnected, etc.
+      // keep in mind that `track.stop()` doesn't trigger this event.
+      track.addEventListener('ended', handleTrackEnded);
+
       transceiver = this.publisher.addTransceiver(track, {
         direction: 'sendonly',
         streams:
           trackType === TrackType.VIDEO || trackType === TrackType.SCREEN_SHARE
             ? [mediaStream]
             : undefined,
-        sendEncodings:
-          trackType === TrackType.VIDEO ? videoEncodings : undefined,
+        sendEncodings: videoEncodings,
       });
 
       this.transceiverRegistry[trackType] = transceiver;
 
-      if (trackType === TrackType.VIDEO) {
-        const codecPreferences = getPreferredCodecs(
-          'video',
-          opts.preferredCodec || 'vp8',
-        );
-
-        if ('setCodecPreferences' in transceiver && codecPreferences) {
-          console.log(`set video codec preferences`, codecPreferences);
-          // @ts-ignore
-          transceiver.setCodecPreferences(codecPreferences);
-        }
-      }
-
-      if (trackType === TrackType.AUDIO) {
-        let returnOnlyMatched = opts.preferredCodec === 'opus';
-        const codecPreferences = getPreferredCodecs(
-          'audio',
-          opts.preferredCodec!,
-          returnOnlyMatched,
+      if ('setCodecPreferences' in transceiver && codecPreferences) {
```

Comment:
```suggestion
      if (Object.hasOwn(transceiver, 'setCodecPreferences') && codecPreferences) {
```

Reply:
`Object.hasOwn` has just been recently introduced in Safari (15.4, which is one year old). Also, this won't work due to some internal differences between `in` and `hasOwn/hasOwnProperty`:

<img width="307" alt="Screenshot 2023-04-27 at 09 35 40" src="https://user-images.githubusercontent.com/843172/234792534-d64bf1db-9247-4bac-9a2b-a5da6e04ad12.png">

Reply:
Oh, TIL - Transceiver object has `setCodecPreferences` somewhere deeper in the prototype. 👍